### PR TITLE
make: introduce $(CLEAN)

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

Currently, some targets are serialized before "clean" by conditionally
adding a dependency.

Make does allow ordering using "|" syntax in prerequisites, but for
"clean" that would always trigger the clean target.
By only conditionally setting $(CLEAN) to clean, that can be
circumvented.

The new CLEAN allows any recipe to be run after "clean" by specifying "|
$(CLEAN)" as prerequisite.

### Testing procedure

Try ```make clean all -j``` and ```make all -j```, ensuring both still work as expected (both with existing and non-existing ```bin/```.

### CODE REVIEW

**Comment by cladmi** [comment]
Makes sense.

Two changes in one commit though.

File: Makefile.include [hunk 1]
```diff
@@ -80,6 +80,13 @@ MAKEOVERRIDES += $(foreach v,$(__DIRECTORY_VARIABLES),$(v)=$($(v)))
 # trailing '/' is important when RIOTPROJECT == CURDIR
 BUILDRELPATH ?= $(patsubst $(RIOTPROJECT)/%,%,$(CURDIR)/)
 
+# Set CLEAN to "clean" if that target was requested.
+# Allowis recipes to be run after cleaning, without triggering it implicitly:
+#
+# all: | $(CLEAN)
+#
+CLEAN=$(filter clean, $(MAKECMDGOALS))
```
Comment (cladmi) [inline Makefile.include]: This is not bash, we use spaces around equals.

Comment (kaspar030) [reply Makefile.include]: yup, fixed

**Comment by kaspar030** [comment]
> Two changes in one commit though.

split in two.

File: Makefile.include [hunk 2]
```diff
@@ -80,6 +80,13 @@ MAKEOVERRIDES += $(foreach v,$(__DIRECTORY_VARIABLES),$(v)=$($(v)))
 # trailing '/' is important when RIOTPROJECT == CURDIR
 BUILDRELPATH ?= $(patsubst $(RIOTPROJECT)/%,%,$(CURDIR)/)
 
+# Set CLEAN to "clean" if that target was requested.
+# Allowis recipes to be run after cleaning, without triggering it implicitly:
```
Comment (fjmolinas) [inline Makefile.include]: *Allows

**Comment by kaspar030** [comment]
> Can you specify that scenario to reproduce and see this fixes it?

Without this fix, ```SUIT_KEY_DIR='$(BINDIR)' make clean all -j``` in #11818 fails (most of the times), as "clean" and "genkey" happens at the same time.

**Comment by fjmolinas** [comment]
@kaspar030 Thanks, verified this fixes the use case in #11818. Please squash directly once typo is fixed.

**Comment by kaspar030** [comment]
> Please squash directly once typo is fixed.

done!

**Comment by cladmi** [comment]
Note for more advanced `|` usages.

This correctly works because the target after `|` is a `.PHONY` target.
When trying to do this with a `| $(BINDIR)/.` target for example and a directory creation rule, the handling was more problematic with the `-j clean all`.

**Pull request merged**